File: snake-strategy-bono/src/main/java/co/electric/snake/strategy/bonostrategy/BonoStrategy.java
package co.electric.snake.strategy.bonostrategy;

import co.electric.snake.framework.model.Arena;
import co.electric.snake.framework.model.Direction;
import co.electric.snake.framework.model.Snake;
import co.electric.snake.framework.strategy.SnakeStrategy;
import co.electric.snake.strategy.bonostrategy.directionprocessor.BlockingDirectionsProcessor;
import co.electric.snake.strategy.bonostrategy.directionprocessor.EquivalentBestDirectionsProcessor;
import co.electric.snake.strategy.bonostrategy.directionprocessor.filtereddirections.FilteredDirectionsProcessor;
import co.electric.snake.strategy.bonostrategy.newdirectionprocessor.NewDirectionProcessorChain;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class BonoStrategy implements SnakeStrategy {

    private static final Logger LOG = LoggerFactory.getLogger(BonoStrategy.class);

    private final FilteredDirectionsProcessor filteredDirectionsProcessor;
    private final EquivalentBestDirectionsProcessor equivalentBestDirectionsProcessor;
    private final BlockingDirectionsProcessor blockingDirectionsProcessor;
    private final NewDirectionProcessorChain newDirectionProcessorChain;

    public BonoStrategy(FilteredDirectionsProcessor filteredDirectionsProcessor, EquivalentBestDirectionsProcessor equivalentBestDirectionsProcessor, BlockingDirectionsProcessor blockingDirectionsProcessor, NewDirectionProcessorChain newDirectionProcessorChain) {
        this.filteredDirectionsProcessor = filteredDirectionsProcessor;
        this.equivalentBestDirectionsProcessor = equivalentBestDirectionsProcessor;
        this.blockingDirectionsProcessor = blockingDirectionsProcessor;
        this.newDirectionProcessorChain = newDirectionProcessorChain;
    }

    @Override
    public Direction nextMove(Snake snake, Arena arena) {
        LOG.info("--- BEGIN " + snake.getName() + " ---");
        LOG.info("Length: " + snake.getLength());
        LOG.info("Food: " + arena.getFoodInNewList().get(0).getCoordinate());
        LOG.info("Head: " + snake.getHeadCoordinate());

        Direction newDirection = process(snake, arena);

        LOG.info("--- END " + snake.getName() + " ---");

        return newDirection;
    }

    private Direction process(Snake snake, Arena arena) {
        SimpleDirectionContainer filteredDirections = filteredDirectionsProcessor.getDirections(snake, arena);
        SimpleDirectionContainer equivalentBestDirections = equivalentBestDirectionsProcessor.getDirections(snake, arena, filteredDirections);
        BlockingDirectionContainer blockingDirections = blockingDirectionsProcessor.getDirections(snake, arena, filteredDirections);

        return newDirectionProcessorChain.process(filteredDirections, equivalentBestDirections, blockingDirections);
    }

}

File: snake-strategy-bono/src/main/kotlin/co/electric/snake/strategy/bonostrategy/BonoStrategy.kt
package co.electric.snake.strategy.bonostrategy

import co.electric.snake.framework.model.Arena
import co.electric.snake.framework.model.Direction
import co.electric.snake.framework.model.Snake
import co.electric.snake.framework.strategy.SnakeStrategy
import co.electric.snake.strategy.bonostrategy.directionprocessor.*
import co.electric.snake.strategy.bonostrategy.directionprocessor.closeddirections.ClosedDirectionsProcessor
import co.electric.snake.strategy.bonostrategy.newdirectionprocessor.NewDirectionProcessorChain
import org.slf4j.LoggerFactory

class BonoStrategy(private val safeDirectionsProcessor: SafeDirectionsProcessor, private val freeDirectionsProcessor: FreeDirectionsProcessor, private val closedDirectionsProcessor: ClosedDirectionsProcessor, private val filteredDirectionsProcessor: FilteredDirectionsProcessor, private val equivalentBestDirectionsProcessor: EquivalentBestDirectionsProcessor, private val blockingDirectionsProcessor: BlockingDirectionsProcessor, private val newDirectionProcessorChain: NewDirectionProcessorChain) : SnakeStrategy {

    companion object {
        private val LOG = LoggerFactory.getLogger(BonoStrategy::class.java)
    }

    override fun nextMove(snake: Snake, arena: Arena): Direction {
        LOG.info("--- BEGIN ${snake.name} ---")
        LOG.info("Head: ${snake.getHeadCoordinate()}")
        LOG.info("Length: ${snake.getLength()}")
        LOG.info("Food: ${arena.getFoodInNewList()[0].coordinate}")
        val newDirection = process(snake, arena)
        LOG.info("--- END ${snake.name} ---")
        return newDirection
    }

    private fun process(snake: Snake, arena: Arena): Direction {
        val unsafeDirections = safeDirectionsProcessor.getDirections(snake, arena)
        val freeDirections = freeDirectionsProcessor.getDirections(snake, arena)
        val closedDirections = closedDirectionsProcessor.getDirections(snake, arena, freeDirections)

        val filteredUnsafeDirections = filteredDirectionsProcessor.getDirections(freeDirections, unsafeDirections)
        val filteredSafeDirections = filteredDirectionsProcessor.getDirections(filteredUnsafeDirections, closedDirections)
        val equivalentSafeBestDirections = equivalentBestDirectionsProcessor.getDirections(snake, arena, filteredSafeDirections)

        val filteredFreeDirections = filteredDirectionsProcessor.getDirections(freeDirections, closedDirections)
        val equivalentFreeBestDirections = equivalentBestDirectionsProcessor.getDirections(snake, arena, filteredFreeDirections)

        val blockingDirections = blockingDirectionsProcessor.getDirections(snake, arena, filteredFreeDirections)

        // val safeDirections = safeDirectionsProcessor.getDirections(snake, arena)
        // val freeDirections = freeDirectionsProcessor.getDirections(snake, arena)
        // val closedDirections = closedDirectionsProcessor.getDirections(snake, arena, freeDirections)
        // val filteredSafeDirections = filteredDirectionsProcessor.getDirections(safeDirections, closedDirections)
        // val equivalentSafeBestDirections = equivalentBestDirectionsProcessor.getDirections(snake, arena, filteredSafeDirections)
        // val filteredFreeDirections = filteredDirectionsProcessor.getDirections(freeDirections, closedDirections)
        // val equivalentFreeBestDirections = equivalentBestDirectionsProcessor.getDirections(snake, arena, filteredFreeDirections)
        // val blockingDirections = blockingDirectionsProcessor.getDirections(snake, arena, filteredFreeDirections)

        return newDirectionProcessorChain.process(filteredSafeDirections, equivalentSafeBestDirections, filteredFreeDirections, equivalentFreeBestDirections, blockingDirections)
    }

}
